### PR TITLE
JDK-8268335: Find better way to exclude empty HTML elements

### PR DESCRIPTION
**Pull request by hns**
Please review a change to straighten out the mechanism to avoid generating empty HTML elements in javadoc. 

Currently this is implemented using the `Content.isValid()` method in `HtmlTree.add(Content)` to check whether the the content argument should be added or dropped. This seems wrong since we don't really want to check for validity but non-emptiness (we want to drop empty elements, not print a warning or error).

So the first decision was to get rid of the `Content.isValid()` method and use `Content.isEmpty()` instead. I considered keeping `isValid()` and using it to actually do some validation checks, but in the end decided it didn't make much sense and scrapped the method.

My initial approach was to take care of emptiness at the content creation site, but it soon turned out there are too many sites potentially creating empty content. After fixing 40+ sites using a new `Content.addNonEmpty(Content)` method I still got empty elements that shouldn't be there in the JDK documentation.

Therefore `HtmlTree.add(Content)` still checks for and drops empty content arguments. Instead, I added a new `HtmlTree.addUnchecked(Content)` method that adds the argument without the checks. The new method is only used 5 times in the whole codebase. It replaces the use of the special `HtmlTree.EMPTY` content token which is no longer needed.

A few remarks about the rewritten `HtmlTree.isEmpty()` method: its purpose is to determine whether content is missing in an element that should have content. It therefore always returns `false` for elements that are expected or allowed to be without content. This is maybe a bit counter-intuitive, but it is what is needed. We could do a combined check such as `isVoid() || !isEmpty()` but that would still leave out elements that *can* be empty, such as `<script>`.

The implementation of `HtmlTree.isEmpty()` is deliberately simple and conservative. For instance, it doesn't look at attributes to decide whether an element is allowed to be empty. The rationale for this is to make the behavior as predictable as possible and avoid surprises.

Since we no longer have a validity check in `HtmlTree.add(Content)` I had to remove a few lines of code in `TestHtmlDocument.java` that expected invalid tags to be dropped. Otherwise there are no test changes, as this is a noreg-cleanup task. I did compare the javadoc JDK documention before and after this change, the output is unchanged.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8268335](https://bugs.openjdk.java.net/browse/JDK-8268335): Find better way to exclude empty HTML elements

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to [fb3bb9da](https://git.openjdk.java.net/jdk/pull/7159/files/fb3bb9da4eadde08b605b0beb9885dc9c0526323)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7159/head:pull/7159` \
`$ git checkout pull/7159`

Update a local copy of the PR: \
`$ git checkout pull/7159` \
`$ git pull https://git.openjdk.java.net/jdk pull/7159/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7159`

View PR using the GUI difftool: \
`$ git pr show -t 7159`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7159.diff">https://git.openjdk.java.net/jdk/pull/7159.diff</a>

</details>

### CODE REVIEW

**Comment by jonathan-gibbons** [comment]
> A few remarks about the rewritten HtmlTree.isEmpty() method: its purpose is to determine whether content is missing in an element that should have content. It therefore always returns false for elements that are expected or allowed to be without content. This is maybe a bit counter-intuitive, but it is what is needed. We could do a combined check such as isVoid() || !isEmpty() but that would still leave out elements that can be empty, such as <script>.

I agree the current situation is not great, and needs revision, so thanks for taking this on.

While it may be hard to completely remove the  creation and use of empty elements, we should at least try and reduce the number.  The current state is a hangover from very-old javadoc, which used print-to-stream before even `HtmlTree`.

Having `isEmpty` return true for some empty elements is definitely counter-intuitive and suggests the name is wrong.
This seems better suited for `isValid` with a definition involving `isVoid` and `isEmpty` as you suggested.  I don't quite understanding the comment about `<script>`. Maybe you are wanting a way to mark elements which should not be considered empty, as in "this space intentionally left blank".  IIRC, that was the original purpose of `HtmlTree.EMPTY` so perhaps some sort of marker element can stay, perhaps with a better name, to mark HTML elements which should not be marked invalid because they appear to be empty.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlTree.java [hunk 1]
```diff
@@ -986,37 +1012,18 @@ public boolean hasAttr(HtmlAttr attrName) {
     }
 
     /**
-     * Returns true if the HTML tree is valid. This check is more specific to
-     * standard doclet and not exactly similar to W3C specifications. But it
-     * ensures HTML validation.
+     * Returns {@code true} if the HTML tree does not affect the output and can be discarded.
+     * This implementation considers non-void elements without content as discardable, with the
+     * exception of {@code SCRIPT} and {@code A} which can sometimes be used without content.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlTree.java:1017]: In HTML 5, when is it reasonable to have `A` without content?

**Comment by hns** [comment]
> 
> There is one question about why `HtmlTree.A` might reasonably be empty in HTML 5. The only reason I can think of is to declare an `id`, since an `href` with no content is unclickable. But a node with just an `id` and no content can apply to other nodes as well, can't it? e.g. `span`

You are right in that this is not correct in context of HTML 5. My purpose was to be liberal in what is accepted. Specifically `TestHtmlDocument.java` contains some HTML constructs which are not quite up to date and I wanted to continue accepting them as far as possible. But it's true that in this case it's mostly the `id` attribute which makes an empty element worth keeping, so I changed the criteria to look for that instead of the `A` tag.

**Review by jonathan-gibbons — APPROVED** [review]
We've maybe not seen the last of this, but this is an improvement on what was there before.

**Pull request closed without merging**